### PR TITLE
feat(#44): Add ability to render multiple errors

### LINKED ISSUE #44: Add ability to render multiple errors in Form Fields
At the moment we only support a string for the `@error` argument; however, we should likely support:

- A single string
- An array of strings, possibly rendered as a list ( `<ul>/<ol>`)
- Yield the error block via `{{yield to="error"}}`

background: https://github.com/CrowdStrike/ember-toucan-core/pull/36#discussion_r1114039128

### PR DESCRIPTION
**Pull request by ynotdraw**
## 🚀 Description
This PR allows for having either a single string for an error or an array of errors.  This will allow us to play more nicely with `ember-headless-form`.

Closes #44 

---

## 🔬 How to Test

- View https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/field#yielded-items and review the `Yielded Items` > `Error` section for validity
- Verify can render both a single error and multiple errors by clicking the link and scroll down a bit (should be the last example). Please also verify the "Error" section of each page
  - https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/fieldset#all-ui-states
  - https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/checkbox-field#all-ui-states
  - https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/textarea-field#all-ui-states
  - https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/input-field#inputfield-with-multiple-errors || https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/input-field#inputfield-with-label-and-error
  - https://aa13bb8f.ember-toucan-core.pages.dev/docs/components/radio-group-field#all-ui-states

---

## 📸 Images/Videos of Functionality

Screenshot 2023-03-20 at 10 26 51 AM

### CODE REVIEW

File: docs/components/checkbox-field/index.md [hunk 1]
```diff
@@ -12,7 +12,18 @@ Provide a string to `@hint` to render the text into the Hint section of the Fiel
...
+
+```hbs
+<Form::CheckboxField @label='Single error' @error='Error' />
+```
+
+```hbs
+<Form::CheckboxField
+  @label='Multiple errors'
+  @error={{(array 'Error 1' 'Error 2')}}
+/>
```
Comment (ynotdraw) [inline docs/components/checkbox-field/index.md:25]: I feel as though `@error` having the option to be both single and plural is not great, but I don't have a better name. I think we definitely want to allow either `string | array<string>` though, so 🤷

File: docs/components/input-field/index.md [hunk 2]
```diff
@@ -79,6 +94,7 @@ Target the error block via `data-error`.
 ## UI States
 
 ### InputField with Label
+
```
Comment (ynotdraw) [inline docs/components/input-field/index.md:97]: Some auto-format prettier markdown stuff due to whitespace - if this annoys you, you can toggle to hide whitespace in GitHub settings! It's only this file though!

File: ember-toucan-core/src/-private/components/error.hbs [hunk 3]
```diff
@@ -18,5 +22,15 @@
     />
   </svg>
 
-  {{yield}}
+  {{#if this.hasMoreThanOneError}}
+    {{! Unfortunately we need this inline style to help with the ul offset  }}
+    {{! template-lint-disable no-inline-styles}}
+    <ul class="m-0 list-none space-y-1 p-0" style="margin-top: -2px">
```
Comment (ynotdraw) [inline ember-toucan-core/src/-private/components/error.hbs:28]: Without the negative-margin you get this 😭 

Screenshot 2023-03-20 at 10 36 22 AM

Comment (ynotdraw) [reply ember-toucan-core/src/-private/components/error.hbs]: Fixed with https://github.com/CrowdStrike/ember-toucan-core/pull/83/commits/11b591e8913907e603276c59ddeb5499eca9913c

File: docs/components/fieldset/index.md [hunk 4]
```diff
@@ -20,12 +20,16 @@ Provide a string to `@hint` to render the text into the Hint section of the Fiel
...
 
-Provide a string to `@error` to render the text into the Error section of the Fieldset. This is optional.
+Provide a string or array of strings to `@error` to render the text into the Error section of the Fieldset. This is optional.
 
 ```hbs
 <Form::Fieldset @label='Label' @error='Error' />
 ```
 
+```hbs
+<Form::Fieldset @label='Label' @error={{(array 'Error 1' 'Error 2')}} />
```
Comment (nicolechung) [inline docs/components/fieldset/index.md:30]: Nice!

File: ember-toucan-core/src/components/form/textarea-field.ts [hunk 5]
```diff
@@ -1,15 +1,16 @@
 import Component from '@glimmer/component';
 import { assert } from '@ember/debug';
 
+import type { ErrorMessage } from '../../-private/types';
```
Comment (nicolechung) [inline ember-toucan-core/src/components/form/textarea-field.ts:4]: Are we exporting / going to export types via an `index` file, or do these types get exported already?

Just thinking this would be useful for folks using the components...

File: test-app/tests/integration/components/field-test.gts [hunk 6]
```diff
@@ -96,4 +96,28 @@ module('Integration | Component | Field', function (hooks) {
     assert.dom('[data-test-hintId]').hasAnyText();
     assert.dom('[data-test-errorId]').hasAnyText();
   });
+
+  test('it renders an array of errors', async function (assert) {
+    let testErrors = ['error 1', 'error 2'];
```
Comment (nicolechung) [inline test-app/tests/integration/components/field-test.gts:101]: Suggestion, non-blocking (maybe typescript will catch this anyhow??): can we test for an array of numbers or a not-string/not-array-of-string condition?

File: ember-toucan-core/src/components/form/input-field.hbs [hunk 7]
```diff
@@ -22,7 +22,7 @@
       />
     </field.Control>
     {{#if @error}}
```
Comment (nicolechung) [inline ember-toucan-core/src/components/form/input-field.hbs:24]: Question: does this `if` statement still display errors if the array happens to be empty?

File: test-app/tests/integration/components/field-test.gts (hunk 6 above)
Comment (nicolechung) [reply test-app/tests/integration/components/field-test.gts:101]: Also: maybe we should test for an empty array here, to confirm that no errors are displayed.

File: docs/components/checkbox-field/index.md (hunk 1 above)
Comment (nicolechung) [reply docs/components/checkbox-field/index.md:25]: Maybe this can be updated to `errors` to match ember-headless-form: https://github.com/CrowdStrike/ember-toucan-core/pull/70/files#diff-9daa8ab5bdba22a48199ecf32b85e898ad08bcff962dc6824c40cc068aca25d5R18

But I think that can be a follow up to make it easier to read the changes.

**Review by nicolechung — APPROVED** [review]

File: test-app/tests/integration/components/field-test.gts (hunk 6 above)
Comment (ynotdraw) [reply test-app/tests/integration/components/field-test.gts:101]: Yeah actually now that I'm thinking about it, I think I have a bug here as an empty array I think will render the error blocks 😬. That's what I get for dropping caffeine and writing code in the AM 😂. Thank you for catching this! Both sound great - I'll add an assertion for the strings of the array as well tomorrow AM!

File: docs/components/checkbox-field/index.md (hunk 1 above)
Comment (simonihmig) [reply docs/components/checkbox-field/index.md:25]: Hm, either way it can be confusing (a little bit at least), like `@errors="Fill in this field"` is also not what you would probably expect. 

I think it's ok the way it is, at least I have no better suggestion...

**Review by simonihmig — APPROVED** [review]

**Comment by ynotdraw** [comment]
Alright, after talking with @nicolechung we are going to merge this as-is knowing that an empty array will force the error styling as to unblock https://github.com/CrowdStrike/ember-toucan-core/pull/70 from merging.

I filed an issue to handle the empty array case at https://github.com/CrowdStrike/ember-toucan-core/issues/90 and will pick this up next!
